include/stdexec/execution.hpp

```diff
@@ -3233,14 +3248,23 @@ namespace stdexec {
           template <class _Self>
           using __completions_t =
             make_completion_signatures<
-              _Sender&,
+              // NOT TO SPEC:
+              // The spec says this should be copy_cvref, but the
+              // 'lvalue split move only sender' test fails if copy_cvref
+              // is used. E.g., in
+              //
+              //   auto snd = just(move_only_type{});
+              //   auto snd2 = split(snd);
+              //
+              // the lvalue 'snd2' does not satisfy the 'sender' concept since
+              // just(move_only_type{}) is move only.
```

Comment:
Now that I think about it, I don't think this test case _should_ work. `snd2` should end up storing (a copy of) `snd` by value. If it's stored by reference, then the split sender will become invalid as soon as the input sender does. But if `snd` is an lvalue to a move-only type, there's no way for `snd2` to store a copy of it.

Reply:
Ah, you are right. I thought that was the simplified version of what I was actually having trouble with, which is getting the following to be valid (the "split into then" test in test_split.cpp)

```
ex::split(ex::just(move_only_type{0}))
    | ex::then([](const move_only_type&) { });
```

I'm not sure if the bug is in the make_completion_signatures alias or get_completion_signatures_t signature, but my current version of make_completion_signatures was *a way* for this expression to compile.

Reply:
Actually, I take it all back. Don't listen to me. `snd2` doesn't need to store a copy of `snd` at all. It stores a `shared_ptr` to the `__sh_state`, in which the _operation state_ lives (and the sender's attrs). So that means that the type of the sender passed to `split` should be perfectly forwarded all the way into `__sh_state`, and _that_ type is used to compute the type of the inner operation state. If a `S&` is passed to `split`, then `__sh_state` should be parameterized on `S&`, and an `S&` is passed to `connect` in `__sh_state`'s c'tor. And that should work even if `S` is a move-only type.

Sorry for the confusion. LMK if I have now thoroughly confused you, and I can take a stab at it.

Reply:
I don't think I'm thoroughly confused :) - except possibly by this:

> If a `S&` is passed to `split`, then `__sh_state` should be parameterized on `S&`, and an `S&` is passed to `connect` in `__sh_state`'s c'tor. And that should work even if `S` is a move-only type.

`split(s)` should not be a valid expression when `s=just(move_only_type{})` - correct? Only `split(move(s))` should be valid, in which case `__sh_state` is parameterized on `S`. Really, I'm interested in single-shot senders, which is the case for `s`.

Reply:
I *think* my latest commit has `split()` behaving as described in the spec after updating the `__sender_t` alias on line 3314 to no longer remove any references from the type . Now, if `S&` is passed to `split`, then the inner sender and shared state templates are parameterized on `S&`, and likewise for `S` being passed to `split`. The completion signatures still doesn't like `copy_cvref_t` however.

Reply:
Looking closer at the spec,

 > The name execution::split denotes a customization point object. For some subexpression s, let S be decltype((s)). If execution::sender_in<S, split-env> ... is false, execution::split is ill-formed. 

In `s=just(move_only_type{})`, `S` will be an lvalue reference, and `sender<S>` is not true since `s` cannot be copied. Is my understanding correct?

In split's completion signatures definition in terms of `make_completion_signatures<
  copy_cvref_t<S2, S>, ...` isn't `split` logically meant to decouple the input sender from the result (hence copy_cvref_t doesn't seem correct here)?

Reply:
I think you're right.